### ISSUE TITLE
Sorting (insertion & merge)

### ISSUE THREAD

**Issue by sstadelman** [sstadelman/sstadelman.github.io#8]
Some problems become easy once the data set is sorted. 

Sorting enables binary search. 

If compressing a file, you could sort items, and maintain count table for duplicates (similar to document distance), so often used as a subroutine.

In graphics rendering, you have many layers, and you're typically rendering front-to-back, so that you can avoid rendering occluded items. Depending on transparency, you may also render back-to-front, so you would typically have a sorted list in both directions.

## Insertion Sort
For `I = 1, 2, ... n`
    insert A[I] into sorted array A[I - 1]
    by pairwise swaps

5, 2, 4, 6, 1, 3
set key (pointer to current location) to 2
since 2 < 5, swap those
2, 5, 4, 6, 1, 3
now key is at position 3 (4)
4 < 5, swap those
2, 4, 5, 6, 1, 3
now key is at position 4 (6)
6 > 5, so no operation
now key is at position 5 (1)
1 < 6, so swap those
1 < 5, so swap those
1 < 4, so swap those
1 < 2, so swap those
1, 2, 4, 5, 6, 3
now key is at position 6 (3)
repeat procedure for 3

movement of the key:
n - 1 steps (since we started at index 2)

moving key item to correct position:
up to n swaps (average n/2 for reverse-ordered list)

`n^2` complexity for swaps & comparisons without optimization 

### interesting optimizations
Take the example though where the comparison itself is much more expensive than swapping register locations, and we want to reduce the complexity of the comparison set.  Since what the algorithm is doing is picking the key value, then trying to insert it at the correct position in the _already sorted subarray_ preceding it, we can use binary search on that subarray to reduce that complexity to `log n` for comparison

## Merge Sort
from array A                                     size N
split into two parts L and R            size N/2
sort into L' and R'                            size N/2 (sorted)
merge L' and R'                                size N

basically no work takes place until we reach the leaves, then call `merge()`

### Merge routine 
takes two sorted arrays as input
(2-finger algorithm -- each finger points to an index in each of the arrays)

creates a new array to merge the sorted arrays

Benefit of Insertion sort vs Merge sort:  insertion does not require auxiliary space.  Merge-sort requires `theta(n)` auxiliary space. (can be reduced to `n/2` by re-using half of the array for copying).  There _is_ a variant of merge sort which is in-place, but it runs much worse than `n log n` so is infrequently used.

'In-place' sort is `theta(1)` auxiliary space.
